src/sentry/workflow_engine/handlers/detector/stateful.py

```diff
@@ -295,15 +297,172 @@ def thresholds(self) -> DetectorThresholds:
+        )
+
+    def _extract_value_from_packet(
+        self,
+        data_packet: DataPacket[DataPacketType],
+    ) -> dict[DetectorGroupKey, DataPacketEvaluationType]:
+        """
+        This method will normalize the extracted value to support grouping results.
+
+        If `extract_value` returns a `dict[DetectorGroupKey, DataPacketEvaluationType]`
+        it will cast it to the correct data type.
+
+        If `extract_value` returns a single value, it will be wrapped in a dict
+        with `None` as the key, to normalize the type as `dict[DetectorGroupKey, DataPacketEvaluationType]`.
+        """
+        data_values = self.extract_value(data_packet)
+        group_data_values: dict[DetectorGroupKey, DataPacketEvaluationType] = {}
+
+        # Normalize the type to dict[DetectorGroupKey, DataPacketEvaluationType]
+        if self._is_detector_group_value(data_values):
+            group_data_values = cast(dict[DetectorGroupKey, DataPacketEvaluationType], data_values)
+        else:
+            group_data_values = {None: cast(DataPacketEvaluationType, data_values)}
+
+        return group_data_values
+
+    def _build_detector_evaluation_result(
+        self,
+        group_key: DetectorGroupKey,
+        new_priority: DetectorPriorityLevel,
+        condition_results: ProcessedDataConditionGroup,
+        data_packet: DataPacket[DataPacketType],
+    ) -> DetectorEvaluationResult:
+        detector_result: IssueOccurrence | StatusChangeMessage
+        event_data: EventData | None = None
+
+        if new_priority == DetectorPriorityLevel.OK:
+            # Call the `create_resolve_message` method to create the status change.
+            detector_result = self._create_resolve_message(group_key)
+            self.state_manager.enqueue_counter_reset(group_key)
+        else:
+            # Call the `create_occurrence` method to create the detector occurrence.
+            detector_occurrence, event_data = self.create_occurrence(
+                condition_results, data_packet, new_priority
+            )
+            detector_result = self._create_decorated_issue_occurrence(
+                detector_occurrence, condition_results, new_priority, group_key
+            )
+
+            # Set the event data with the necessary fields
+            event_data["timestamp"] = detector_result.detection_time
+            event_data["project_id"] = detector_result.project_id
+            event_data["event_id"] = detector_result.event_id
+            event_data.setdefault("platform", "python")
+            event_data.setdefault("received", detector_result.detection_time)
+            event_data.setdefault("tags", {})
+
+        return DetectorEvaluationResult(
+            group_key=group_key,
+            is_triggered=new_priority != DetectorPriorityLevel.OK,
+            priority=new_priority,
+            result=detector_result,
+            event_data=event_data,
+        )
+
+    def _is_detector_group_value(self, value) -> bool:
```

Comment:
is there another way i can tell if this is a `DataPacketEvaluationType` or `dict[DetectorGroupKey, DataPacketEvaluationType]`? (i can't use `isinstance` on `DataPacketEvaluationType` because it's generic 😭)